xarray/core/datatree_mapping.py

```diff
@@ -1,236 +1,129 @@
 
 if TYPE_CHECKING:
     from xarray.core.datatree import DataTree
 
 
-class TreeIsomorphismError(ValueError):
-    """Error raised if two tree objects do not share the same node structure."""
+@overload
+def map_over_datasets(func: Callable[..., Dataset | None], *args: Any) -> DataTree: ...
 
-    pass
 
+@overload
+def map_over_datasets(
+    func: Callable[..., tuple[Dataset | None, Dataset | None]], *args: Any
+) -> tuple[DataTree, DataTree]: ...
 
-def check_isomorphic(
-    a: DataTree,
-    b: DataTree,
-    require_names_equal: bool = False,
-    check_from_root: bool = True,
-):
-    """
-    Check that two trees have the same structure, raising an error if not.
-
-    Does not compare the actual data in the nodes.
 
-    By default this function only checks that subtrees are isomorphic, not the entire tree above (if it exists).
-    Can instead optionally check the entire trees starting from the root, which will ensure all
+@overload
+def map_over_datasets(
+    func: Callable[..., tuple[Dataset | None, ...]], *args: Any
+) -> tuple[DataTree, ...]: ...
 
-    Can optionally check if corresponding nodes should have the same name.
 
-    Parameters
-    ----------
-    a : DataTree
-    b : DataTree
-    require_names_equal : Bool
-        Whether or not to also check that each node has the same name as its counterpart.
-    check_from_root : Bool
-        Whether or not to first traverse to the root of the trees before checking for isomorphism.
-        If a & b have no parents then this has no effect.
-
-    Raises
-    ------
-    TypeError
-        If either a or b are not tree objects.
-    TreeIsomorphismError
-        If a and b are tree objects, but are not isomorphic to one another.
-        Also optionally raised if their structure is isomorphic, but the names of any two
-        respective nodes are not equal.
+def map_over_datasets(
+    func: Callable[..., Dataset | None | tuple[Dataset | None, ...]], *args: Any
+) -> DataTree | tuple[DataTree, ...]:
     """
-    # TODO: remove require_names_equal and check_from_root. Instead, check that
-    # all child nodes match, in any order, which will suffice once
-    # map_over_datasets switches to use zip_subtrees.
-
-    if not isinstance(a, TreeNode):
-        raise TypeError(f"Argument `a` is not a tree, it is of type {type(a)}")
-    if not isinstance(b, TreeNode):
-        raise TypeError(f"Argument `b` is not a tree, it is of type {type(b)}")
+    Decorator which turns a function which acts on (and returns) Datasets into one
+    which acts on and returns DataTrees.
 
-    if check_from_root:
-        a = a.root
-        b = b.root
+    Applies a function to every dataset in one or more DataTree objects with
+    the same structure (ie.., that are isomorphic), returning new trees which
+    store the results.
 
-    diff = diff_treestructure(a, b, require_names_equal=require_names_equal)
+    The function will be applied to any data-containing dataset stored in any of
```

Comment:
I don't think you're doing `if node.has_data` any more?
```suggestion
    The function will be applied to any dataset stored in any of
```